File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        if len(nums)==1:
            return TreeNode(nums[0])

        mid = len(nums)//2
        root = TreeNode(nums[mid])

        left_subtree = nums[:mid]
        right_subtree = nums[mid+1:]

        if left_subtree:
            root.left = self.sortedArrayToBST(left_subtree)
        if right_subtree:
            root.right = self.sortedArrayToBST(right_subtree)
        
        return root

File: 0108-convert-sorted-array-to-binary-search-tree/0108-convert-sorted-array-to-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        int mid = nums.size()/2;
        TreeNode* root = new TreeNode(nums[mid]);
        TreeNode* temp1 = root;
        TreeNode* temp2 = root;
        
        for (int i=mid-1; i>=0; i--) {
            TreeNode* node = new TreeNode(nums[i]);
            temp1->left = node;
            temp1 = node;
        }

        for (int j=mid+1; j<nums.size(); j++) {
            TreeNode* node = new TreeNode(nums[j]);
            temp2->right = node;
            temp2 = node;
        }

        return root;
    }
};